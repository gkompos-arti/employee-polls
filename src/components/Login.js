import React, { useState } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { handleLogin } from "../actions/shared";
import { useNavigate } from "react-router-dom";
import loginIntro from '../photos/intro.png';
import { Box, Button, FormControl, InputLabel, MenuItem, Paper, Select } from "@mui/material";
import "./App.css";

function Login(props) {
  const [selectedUser, setSelectedUser] = useState("");
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    const { dispatch } = props;
    dispatch(handleLogin(selectedUser));
    navigate('/dashboard');
  };

  const { users } = props;

  return (
    <Box
      sx={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        height: "100vh",
      }}
    >
      <Paper sx={{ padding: "32px", maxWidth: "500px", width: "100%" }}>
        <Box sx={{ textAlign: "center" }}>
          <h1 data-testid="login-heading">Employee Polls</h1>
          <div className="login-image-container">
            <img src={loginIntro} alt="login intro" className="login-image"/>
          </div>
        </Box>
        <Box sx={{ textAlign: "center", mt: "32px" }}>
          <h1 data-testid="loginPage">Log In</h1>
          <FormControl sx={{ m: 1, minWidth: 240 }}>
            <label id="login-users" data-testid="username">User</label>
            <select value={selectedUser} onChange={(e) => setSelectedUser(e.target.value)} label="User" labelid="login-users" data-testid="select-user">
              <option value="" disabled>
              Select User
              </option>
              {Object.keys(users).map((userId) => (
                <option key={userId} value={userId} data-testid="select-user-item">
                  {users[userId].name}
                </option>
              ))}
            </select>
            <Button  className="loginBtn" name="login" type="submit" disabled={selectedUser === ""} sx={{marginTop: "15px"}} 
            variant="outlined" color="secondary" onClick={(e) => {handleSubmit(e)}} data-testid="login">
              Login
            </Button>
          </FormControl>
        </Box>
      </Paper>
    </Box>
  );
}

function mapStateToProps({ users }) {
  return {
    users,
  };
}

export default connect(mapStateToProps)(Login);