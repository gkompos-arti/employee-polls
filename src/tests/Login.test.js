import {fireEvent, render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../store";
import {BrowserRouter} from "react-router-dom";
import '@testing-library/jest-dom';
import React from "react";
import Login from "../components/Login";


describe("Login", () => {
    it("should render the component", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('renders the dropdown and submit button', () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        // Get the dropdown and submit button by their data-testid attributes
        const dropdown = component.getByTestId('username');
        const submitButton = component.getByTestId('login');
    
        // Assert that the dropdown and submit button are both in the document
        expect(dropdown).toBeInTheDocument();
        expect(submitButton).toBeInTheDocument();
      });

      it('the login button must unlock', () => {

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        
        const loginHeadingElement = wrapper.getByTestId("loginPage");
        const userSelectElement = wrapper.getByTestId("select-user");
        // const select = userSelectElement.childNodes[1]
        const loginButtonElement = wrapper.getByTestId("login");


        expect(loginButtonElement).toBeInTheDocument();
        expect(userSelectElement).toBeInTheDocument();
        expect(loginButtonElement).toBeInTheDocument();

        fireEvent.select(userSelectElement, {target: {value: "sarahedo"}});
        
        expect(loginButtonElement).toBeDisabled(false);
    });
});