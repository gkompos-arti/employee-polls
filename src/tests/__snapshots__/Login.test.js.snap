// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1lb6n3v"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1edzwbn-MuiPaper-root"
        >
          <div
            class="MuiBox-root css-xi606m"
          >
            <h1
              data-testid="login-heading"
            >
              Employee Polls
            </h1>
            <div
              class="login-image-container"
            >
              <img
                alt="login intro"
                class="login-image"
                src="intro.png"
              />
            </div>
          </div>
          <div
            class="MuiBox-root css-4oqe9z"
          >
            <h1
              data-testid="loginPage"
            >
              Log In
            </h1>
            <div
              class="MuiFormControl-root css-1kzk0u5-MuiFormControl-root"
            >
              <label
                data-testid="username"
                id="login-users"
              >
                User
              </label>
              <select
                data-testid="select-user"
                label="User"
                labelid="login-users"
              >
                <option
                  disabled=""
                  value=""
                >
                  Select User
                </option>
              </select>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium loginBtn css-1tixegr-MuiButtonBase-root-MuiButton-root"
                data-testid="login"
                disabled=""
                name="login"
                tabindex="-1"
                type="submit"
              >
                Login
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1lb6n3v"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1edzwbn-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-xi606m"
        >
          <h1
            data-testid="login-heading"
          >
            Employee Polls
          </h1>
          <div
            class="login-image-container"
          >
            <img
              alt="login intro"
              class="login-image"
              src="intro.png"
            />
          </div>
        </div>
        <div
          class="MuiBox-root css-4oqe9z"
        >
          <h1
            data-testid="loginPage"
          >
            Log In
          </h1>
          <div
            class="MuiFormControl-root css-1kzk0u5-MuiFormControl-root"
          >
            <label
              data-testid="username"
              id="login-users"
            >
              User
            </label>
            <select
              data-testid="select-user"
              label="User"
              labelid="login-users"
            >
              <option
                disabled=""
                value=""
              >
                Select User
              </option>
            </select>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium loginBtn css-1tixegr-MuiButtonBase-root-MuiButton-root"
              data-testid="login"
              disabled=""
              name="login"
              tabindex="-1"
              type="submit"
            >
              Login
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
